### ISSUE TITLE
WebAssembly standard

### ISSUE THREAD

**Issue by carllaw6885** [OWASP/ASVS#708]
Following the comment of a participant at NDC London, please consider a standard for WebAssembly

**Comment by MarcinHoppe** [comment]
What would such a WebAssembly standard entail?

**Comment by carllaw6885** [comment]
Good question and I’m not going to pretend to know. I was in a talk with Jim Manico where some asked this question at the end and he asked for someone to raise an issue on here to answer it.

**Comment by jmanico** [comment]
Thank you for this! We’ll address it somehow in the next version. We’re working on it next month so you’ll see activity then.

Thanks again!
--
Jim Manico
@Manicode
Secure Coding Education
+1 (808) 652-3805

**Comment by jmanico** [comment]
Yea we’re doing a project summit in Cancun next month. We’ll discuss it and research it in detail then!

--
Jim Manico
@Manicode
Secure Coding Education
+1 (808) 652-3805

**Comment by jmanico** [comment]
I have not made progress on this and am new to the technology. Anyone else have any ideas here?

**Comment by sanmai-NL** [comment]
@jmanico A WebAssembly runtime’s [security model](https://webassembly.org/docs/security/) is very particular. Some offensively oriented papers have been written on it, e.g. [Everything Old is New Again: Binary Security of WebAssembly](https://www.usenix.org/conference/usenixsecurity20/presentation/lehmann). ASVS may mostly focus on verification that when converting to or otherwise adding WebAssembly modules, this model is taken into account.

**Comment by jmanico** [comment]
Would you like to propose a few requirements for this section which we can add to the current version?

**Comment by sanmai-NL** [comment]
@jmanico 
ASVS Level 2, V1.13 API Architectural Requirements

> Verify that all applicable security models (e.g., WebAssembly, Same Origin Policy) have been explicitly accounted for, per (sub)system, in the architectural design, at least the interactions between domains that are governed by different models.

**Comment by danielcuthbert** [comment]
We are going to close this issue for now. Please reopen it with your recommendation if you are still interested in working on this.

**Comment by tghosth** [comment]
I think we need to keep this topic and see if we can find anyone else to provide some ideas, I don't think we can ignore it for 5.0
.

**Comment by tghosth** [comment]
This has not had activity recently and will need someone from the community to provide more specific recommendations in order to progress.

**Comment by tghosth** [comment]
https://thenewstack.io/6-security-risks-to-consider-with-webassembly/?utm_content=161073144&utm_medium=social&utm_source=twitter&hss_channel=tw-1462816075490500622

**Comment by tghosth** [comment]
https://thenewstack.io/6-security-risks-to-consider-with-webassembly/

**Comment by tghosth** [comment]
hi @nielstanis any ideas on this?

**Comment by randomstuff** [comment]
Design your WebAssembly module in order to avoid exposing functionalities which are not strictly necessary to the WebAssembly module?

This is not very actionnable, though.

**Comment by jmanico** [comment]
Here s my list:

- Verify that all input data passed to WebAssembly modules is thoroughly validated to prevent injection attacks.
- Verify that WebAssembly modules carefully manage memory usage to avoid out-of-bounds memory access and other memory-related vulnerabilities.
- Verify that WebAssembly modules are isolated from the host environment, are run in a sandbox and are granted only the minimum necessary permissions.
- Verify that all communication between WebAssembly modules and the host environment is encrypted and authenticated.

**Comment by jmanico** [comment]
Regarding sandboxing, non-browser environments may benefit from secure WebAssembly runtimes like https://github.com/bytecodealliance/wasmtime - very similar to what is built into most browsers.

**Comment by jmanico** [comment]
Would you like to propose a PR @randomstuff? I think you get the big picture very well.

**Comment by randomstuff** [comment]
I can provide a first draft. Do we intend to have a dedicated chapter for this? V54 WebAssembly?

**Comment by randomstuff** [comment]
I feel that a dedicated WebAssembly chapter would be quite short. On the other hand I'm not sure where to move these.

Alternatively, the different requirements might be spread out in the different chapter but I don't think that's desirable.

**Comment by tghosth** [comment]
Do we have enough requirements that are specific to web assembly to make it worthwhile to have a section/chapter for this? Most of the requirements I see above apply to other types of app as well.

**Comment by randomstuff** [comment]
> Most of the requirements I see above apply to other types of app as well.

Yes, I agree.

To recap what have been have listed here (with some tweaks).

> Verify that all input data passed to WebAssembly modules is thoroughly validated to prevent injection attacks.

I am wondering if this is much of an issue to deserve a special mention. I would be more wary about data passed by WebAssembly module to the host (which may compromised by some memory corruption in the WebAssembly memory).

> Verify that WebAssembly modules carefully manage memory usage to avoid out-of-bounds memory access and other memory-related vulnerabilities.

This relates to some requirements we have removed in #2280. I would better have a general requirement about memory safety in general and mention WebAssembly modules as an important case.

Anyway, in practice people are going to want to run large pre-existing code base in WebAssembly (think: [Unity](https://docs.unity3d.com/6000.0/Documentation/Manual/wasm-2023-features.html)) and it might be difficult to apply this in practice.

> Verify that all communication between WebAssembly modules and the host environment is encrypted and authenticated.

As discussed earlier I would not include this because it is already covered elsewhere.

> Verify that WebAssembly modules are isolated from the host environment application by the WebAssembly engine. (This includes for example mitigations for Spectre-like vulnerabilities if applicable.)

This could be generalized to other interpreters (eBPF?). This someone amounts to "use a decent WebAssembly implementation", however …

> Verify that the host environment only exposes functionalities to the WebAssembly module which are strictly required for its operation.

For me, this is probably the most important thing we can say about WebAssembly. But then again, this is applicable to other contexts.

So I don't think if we have enough material to have a dedicated section :)

**Comment by jmanico** [comment]
So @randomstuff do think we should close this out and just skip WebAssembly requirements?

**Comment by randomstuff** [comment]
Yes, I think we can close this as we don't have much material for a WebAssembly chapter. The question is whether we might want to include some reference to WebAssembly in existing requirements (or chapters).

Might be relevant to revive #2280 into another form? Something like:

> Verify that memory-unsafe code is identified and that a strategy for mitigating binary exploitation is documented according to a risk analysis. This might include hardening compiler flags, isolation of unsafe modules, etc.

About this one:

> Verify that all input data passed to WebAssembly modules is thoroughly validated to prevent injection attacks.

This might be reformulated in the more general case as something like (not a wording proposal):

> Verify that when data flows from a module to a higher-trust module, all input data is thoroughly validated. (For example: untrusted client→trusted server, WebAssembly module→host application, sandboxed environnement executing user code→host application)

Is there such a requirement?  Would there be a need for one?

Ans this one would be related as well to this consideration:

> Verify that the host environment only exposes functionalities to the WebAssembly module which are strictly required for its operation.
